adr/0007-package-variables-files-revisit.md

```diff
@@ -0,0 +1,135 @@
+
+## Options
+
+With the above context below are options to address the concerns for both `prompt` and `default` keeping in mind that file variables will be added.
+
+### Default
+
+The default confusion could be addressed in one of the following ways:
+
+1. Remove the `default` key altogether and rely on package authors to create example overrides files.
+   - *Pros*: This makes things simpler and allows us not to have to worry about abuses of this key.
+   - *Cons*: This removes options from package authors where they may want to set a legitimate default.  Overrides files also don't natively travel along with the package.
+1. Allow the `default` key to be unset if it does not exist (no implicit "")
+   - *Pros*: This allows package authors to set defaults that travel inside of a package and can be potentially carried into generated example override files.
+   - *Cons*: This would require the use of a pointer in the package types definitions.
+1. Keep the current `default` behavior and force an implicit ""
+   - *Pros*: This is not a breaking change and allows us to not have to have a pointer in our type definitions
+   - *Cons*: This forces an implicit default onto package creators.
+
+### Prompt
+
+The prompt confusion could be addressed in one of the following ways: (Note that these largely depend on the outcome of [Default](#default))
+
+1. Keep the current `--confirm` behavior but detect if there is a default set and error if both `prompt` and `--confirm` are used with no default or `--set` existing for a particular variable
+   - *Pros*: This keeps things simpler with less to type out at the command line.
+   - *Cons*: This is less explicit and could be confusing.  It also depends on implementing a `default` pointer.
+1. Introduce a second flag (i.e. `--accept-defaults`) and error out if a user does not use the flag or set all variables through a file or at the command line
+   - *Pros*: This is more explicit and could work with the latter two `default` COAs.
+   - *Cons*: This requires another command line flag that may just get in the way.
+1. Remove the `prompt` key and detect when to prompt based on `default` existing
+   - *Pros*: This makes things simpler overall
+   - *Cons*: This relies on implementing a `default` pointer and gives package authors less control
+1. Remove the `prompt` key and always prompt
+   - *Pros*: This makes things simpler overall
+   - *Cons*: This could make things unweildy in large packages and gives package authors less control
+
+> Note that for the purposes of this ADR it was decided to exclude considering options making use of the `description` or `matches` fields as those ultimately solve different problems and would only be tangentially related to the above.
+
+## Decision
+
+It was decided to allow the `default` key to be nillable to detect its existence and error out if `--confirm` and `prompt` are used together with no existing default or `--set`.  This would result in the following behavior:
```

Comment:
I still disagree with this conclusion, but whatever we decide as a team, I think it would be good to document who agrees and who disagrees in this ADR and any dissenting comments. Regardless, this is a team decision, not mine.

I've also thought about this some more and might have another option. Can you clarify how this might change if we had a "validation" key as well? I think a viable option is to leave the current behavior as-is and provide a regex matcher so the package author can state what is acceptable. This way, in the case of the implied nil being used to force a failure, instead the failure can occur by setting the validation as requiring some content vs an empty string. 

For the first concern with the current implementation, allowing package authors to explicitly state what is a valid value will enable them to ensure the most important outcome: that an acceptable value was provided for the variable. Forcing this into nuance of whether the default key is specified by the package author seems far less obvious to me if I don't understand the details of yaml or golang typing.

For the second concern, the purpose of prompt really is to serve as a way to expose the variable as a choice for users in interactive mode. This is important because a persona only using interactive mode would have no way to set a variable that isn't set to prompt without using a flag. Prompt gives package authors a way to surface an important variable the user really should look at or change. 

Here's how I see the current and potential future keys working:

```yaml
name: the identifier for the variable
description: (not implemented yet), a message to explain to users the purpose of the variable or additional info to help decide what to enter
default: a value to suggest for users in interactive mode or the default value with --confirm; it's presence or absence shouldn't have nuanced meaning, it should just be a default value if provided, no other side effects
prompt: give users in interactive mode the ability to change the variable during package deploy, no meaning for --confirm as the key name itself requires interaction; this could have been interactivePrompt or similar I guess, but that would be longer
validation: (not implemented yet), regex to match whatever conditions the package author needs, including "not an empty string" 
```

Thoughts?

Reply:
Definitely want @YrrepNoj and @RothAndrew to weigh in too but I don't really see that `default` behavior as a nuance but rather as being more clear.  IMO if an author doesn't provide a default they shouldn't get a default (regardless of their knowledge of yaml/golang).  I agree validator or match would be a good thing to implement down the road, but IMO it is a bandaid to patch over on top of this problem.

Reply:
I agree with @Racer159 on all points. If an author didn't provide a default Zarf should not inject one behind the scenes.

Reply:
I don't think validation is a bandaid, I think it's the right eventual solution in all cases. But either way, please update the ADR with this proposed option, even if the team doesn't select it, it needs to be listed as an option.

Reply:
I agree that it should be implemented regardless (and will definitely document it as a COA for these issues).  I'm also gonna add some logic images to this too to help break things down.  This is how I see validation alone working:

![Validation Behavior drawio](https://user-images.githubusercontent.com/3977569/186732712-a976d811-8dda-42b4-9f88-6309fd8853f8.svg)

Reply:
So two things, --confirm can never have additional user prompting. We currently break anything that is missing/incorrect if --confirm is provided. Reason for this is --confirm is used for automation and needs to fail vs just hang for an indefinite amount of time.

What I'm describing is exactly this flow without the "intentional default" box. I maintain it's nuance and implicit and lacks specificity for perceived convenience at the risk of further confusion. Granted, in this case I'm the only one of us 4 that sees it this way I think. But in the particular case that someone needs to have a value present for the template string to be valid, a valuation key is far more robust than "did the author remember to put a default key before creating the packaging and shipping it off."  

Across the 58 sting declarations in zarf types, we do not have a single case of key omission being nil. However, we are able to leverage json schema required for most of those keys. I think the issue at hand is we don't know if required is actually true or not.  At this point I'd rather just add "required" or "noEmpty" or something than use the presence or lack of a default key to mean that.  Though I maintain that validation also does this and more too, and would take half an hour to implement.

Reply:
--confirm skips prompting in the above example, I only made the Fail purple to distinguish an intentional failure from an unintentional one.

Without detecting the presence of a default key how do you know if the default was intentional and remove that logic point?

Also, `noEmpty` would also fail if an author intentionally wanted an empty default (i.e. a license file).

Reply:
yeah sorry I updated the comment, I misread the --confirm flow.

Reply:
In the license key example, where empty is valid, you wouldn't set `noEmpty` to true, or you wouldn't add additional validation because nothing is already valid.  I do not prefer `require` or `noEmpty` at all, I think that `validation` is far more robust and clear. I was only suggesting as a point of illustration how that could work, though not preferred.  

I think we all agree validation is useful and powerful.  I also think we need to just assume it exists now, and I can make the PR for it this afternoon if that's easier. My question then is, how are these concerns not addressed with validation in place.

Reply:
Ah ok I read that as being jsonschema rules.  I would argue they are not addressed (hence bandaid) because validation is only optional and still leaves the door open for the implicit default of `empty string`

Reply:
@YrrepNoj and @RothAndrew though for their thoughts

Reply:
> how are these concerns not addressed with validation in place.

Adding a validation will be an intentional action taken by the package author. If they don't add one, because they forgot, or they don't know it exists, the "unintentional default" logic branch will still be there.

The only way of removing that "unintentional default" logic branch while keeping the "nullability" of that field out, is to make the default validation something like `.+`.

Reply:
IMO `.+` as a validation default, while allowing us to avoid `nil`s, is less intuitive to package authors (most validators that exist default to `.*`)

Reply:
That's a good point @RothAndrew, it's really hard for us right now to assume either way, whether on the utility/effectiveness of default or what the default  validation rule would be.  We are basing all of this off of very little data and relying instead on our current contexts to make assumptions about what is logical.  I know we're trying to get to majority / norm practice and doing that without user feedback/data is quite hard. We probably just need to solicit feedback from various user groups before moving forward, both package creators and deployers.  In particular I want to make sure we capture the low-level IT admin persona in interviews.  This sounds like a @Madeline-UX thing.

I honestly don't have a good answer right now on how often variables will actually be acceptable to empty strings or not. For example, if a package author is setting an optional license key, maybe an empty string is totally acceptable.  But they could also be doing something crazy (and we have no way to stop them) like defining custom yaml in the variable.  I don't think that will be the typical use-case, but it is possible I suppose.

Reply:
As a current user of Zarf, package creator and deployer (DI2-ME). I highly dislike the idea of having to remember to add a validation to every field that can't be an empty string, which will be ~99% of the variables that I create.

Reply:
That's useful feedback, though I wouldn't put you in the deployer persona as much as expert package creator.  On DI2-ME I have a separate concern with exposing too many variables to the user on package deploy and think we should dig into that more separately.  We could always expose the "default variable validation" in the schema somewhere, though I have no idea where.

Reply:
Making the validation field required could work as well to eliminate the "unintentional default". I'd prefer that over it being optional if the fall-through case is that the empty string becomes the implicit default.

Reply:
Side note: I propose `match:` as the name of the field. It gives better indication as to what the expected value is in the absence of any documentation.

Reply:
Yeah that is another reason why (at least to me) I like keeping these things discrete `default` | `prompt` | `match`.  IMO if we try to solve one with another that introduces a lot of confusion.

Reply:
And do agree it is worth validating this ADRs flows with @Madeline-UX

Reply:
> That's a good point @RothAndrew, it's really hard for us right now to assume either way, whether on the utility/effectiveness of default or what the default validation rule would be. We are basing all of this off of very little data and relying instead on our current contexts to make assumptions about what is logical. I know we're trying to get to majority / norm practice and doing that without user feedback/data is quite hard. We probably just need to solicit feedback from various user groups before moving forward, both package creators and deployers. In particular I want to make sure we capture the low-level IT admin persona in interviews. This sounds like a @Madeline-UX thing.
> 
> I honestly don't have a good answer right now on how often variables will actually be acceptable to empty strings or not. For example, if a package author is setting an optional license key, maybe an empty string is totally acceptable. But they could also be doing something crazy (and we have no way to stop them) like defining custom yaml in the variable. I don't think that will be the typical use-case, but it is possible I suppose.

@jeff-mccoy I agree that this seems to be an area with a lot of both risk and assumptions and needs more validation.  I Just may need some technical support to come up with the best way to elicit the information we need to answer this question.

Reply:
@jeff-mccoy If we have no validation for the need for nil at this time. How difficult would it be to build without it and then add in at a later point if validation shows that it is needed?